File: WorldSpawnSpigot/src/main/java/me/ifydev/worldspawn/spigot/WorldSpawnMain.java
package me.ifydev.worldspawn.spigot;

import me.ifydev.worldspawn.api.WorldSpawnAPI;
import me.ifydev.worldspawn.api.backend.ConnectionInformation;
import me.ifydev.worldspawn.spigot.utils.ConfigVerifier;
import me.ifydev.worldspawn.spigot.utils.SpigotDisplayManager;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.Optional;

/**
 * @author Innectic
 * @since 08/07/2018
 */
public class WorldSpawnMain extends JavaPlugin {

    private WorldSpawnAPI api;

    @Override
    public void onEnable() {
        Optional<ConnectionInformation> connectionInformation = ConfigVerifier.getConnectionInformation();
        if (!connectionInformation.isPresent()) {
            getLogger().severe("Could not find connection information in configuration. Please verify the options.");
            return;
        }

        api = new WorldSpawnAPI(connectionInformation.get(), new SpigotDisplayManager());
    }

    @Override
    public void onDisable() {
    }

    public static WorldSpawnMain getInstance() {
        return WorldSpawnMain.getPlugin(WorldSpawnMain.class);
    }

    public Optional<WorldSpawnAPI> getAPI() {
        return Optional.ofNullable(api);
    }
}

File: Spigot/src/main/java/me/ifydev/worldspawn/spigot/WorldSpawnMain.kt
package me.ifydev.worldspawn.spigot

import me.ifydev.worldspawn.api.WorldSpawnAPI
import me.ifydev.worldspawn.api.library.LibraryHandler
import me.ifydev.worldspawn.spigot.command.WorldSpawnCommand
import me.ifydev.worldspawn.spigot.events.PlayerSwitchWorldEvent
import me.ifydev.worldspawn.spigot.util.verifyConnectionInformation
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.plugin.java.JavaPlugin

/**
 * @author Innectic
 * @since 08/07/2018
 */
fun String.convertColors(): String {
    return ChatColor.translateAlternateColorCodes('&', this)
}

class WorldSpawnMain : JavaPlugin() {

    lateinit var api: WorldSpawnAPI

    override fun onEnable() {
        LibraryHandler().loadLibraries(logger)

        var initTime = System.currentTimeMillis()

        api = WorldSpawnAPI()
        api.initialize(verifyConnectionInformation(), { permission, messages ->
            Bukkit.getOnlinePlayers().filter { player -> player.hasPermission(permission) }.forEach {
                player -> messages.map {msg -> msg.convertColors() }.forEach { msg -> player.sendMessage(msg) } } })

        initTime = System.currentTimeMillis() - initTime
        logger.info("Initialized WorldSpawn API in ${initTime / 1000} ($initTime ms) seconds!")

        registerListeners()
        registerCommands()
    }

    override fun onDisable() {

    }

    private fun registerListeners() {
        Bukkit.getPluginManager().registerEvents(PlayerSwitchWorldEvent(), this)
    }

    private fun registerCommands() {
        getCommand("worldspawn").executor = WorldSpawnCommand()
    }

    companion object {
        fun getInstance(): WorldSpawnMain = getPlugin(WorldSpawnMain::class.java)
    }
}